Ctv 3359 uat automation

addisonm opened the pull request:
### JIRA
https://infillion.atlassian.net/browse/CTV-3359

### Description
Introduce jenkins UAT automation

### Testing
A few test deployments

### Commit Message Subject(s)
LIST_PROPOSED_COMMIT_SUBJECT_LINES

### Additional Context
OPTIONAL_ADDITIONAL_INFO

### Related PRs
OPTIONAL_ADD_PR_LINKS

### Checks
- [ ] Includes unit test coverage _(if applicable)_
- [ ] Includes functional test coverage _(at feature-level, if applicable)_
- [ ] Proposed commit messages follow [team conventions](https://github.com/socialvibe/adlabs-wiki/blob/develop/git/README.md#commits)

addisonm committed:
CTV-3359
- Iniital commit for uat automation

addisonm committed:
CTV-3359
- Update version in preparation for new tag
- Remove test endpoint, and use expected endpoint

addisonm committed:
CTV-3359
- Changelog

addisonm commented on src/deploy/jenkins-uat.js:
```diff
@@ -0,0 +1,22 @@
+            "skylineBranchName": "develop",
+            "skylineS3BranchDirName": "develop",
+            "fromBuildJobUrl": process.env.TRAVIS_BUILD_WEB_URL,
+            "sourceRepo": process.env.TRAVIS_REPO_SLUG
```
This will look something like `socialvibe/Skyline`

addisonm commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "truex-shared",
-  "version": "1.6.13",
+  "version": "1.7.1",
```
I haven't made the 1.7.1 tag yet, figure I'd wait a bit given the current state

imann24 replied:
Good call! And thanks for handling the bump up to `1.7`. We've been meaning to do that for a bit

imann24 commented on src/deploy/jenkins-uat.js:
```diff
@@ -0,0 +1,22 @@
+const fetch = require('node-fetch');
+
+module.exports = () => {
+    const jenkinsUrl = "https://ctv-jenkins-02.truex.com/generic-webhook-trigger/invoke";
```
Can we allow for a env variable overload here? 
```javascript
const jenkinsUrl = process.env.JENKINS_URL || "https://ctv-jenkins-02.truex.com/generic-webhook-trigger/invoke";
```

rblaa replied:
this is a good idea

addisonm replied:
The overloads don't seem too useful right now, if this is only going to be executed by the travis automation (and currently only does the defaults anyway.  But I can understand the sentiment of making it more future proof.

I wouldn't want to use environment variables for this though, and instead make it parameterized.

imann24 replied:
For the Jenkins URL, it feels worth having a way to change it without requiring a code deploy (especially in this shared package given how many others it touches). Mainly based on the fact the Jenkins URL has changed in the past

imann24 commented on src/deploy/jenkins-uat.js:
```diff
@@ -0,0 +1,22 @@
+            "token": "TAR-HTML5_UAT",
+        },
+        body: JSON.stringify({
+            "whakapapaBranchName": "develop",
```
These may be worth environment variable overrides too

rblaa replied:
this needs to be parameterized, probably make an arg to the function, since the host repo knows the travis settings, not this shared function as such

imann24 commented on src/deploy/jenkins-uat.js:
```diff
@@ -0,0 +1,22 @@
+            "devicePlatforms": "Unknown,Tizen,LG,XboxOne,FireTV,AndroidTV,Vizio,Comcast",
+            "skylineBranchName": "develop",
+            "skylineS3BranchDirName": "develop",
+            "fromBuildJobUrl": process.env.TRAVIS_BUILD_WEB_URL,
```
Nice touch!

imann24 reviewed (approved):
Looks good to me! Allowing more overrides may be handy

rblaa commented on src/deploy/jenkins-uat.js:
```diff
@@ -0,0 +1,22 @@
+const fetch = require('node-fetch');
+
+module.exports = () => {
```
can we have a named function instead, so it is really clear what this is for, e.g.
```
// Inform Jenkins server to start an automation test run for the current branch.
function invokeAutomationTests() {
...
}
module.exports = invokeAutomationTests;

rblaa commented on src/deploy/jenkins-uat.js:
```diff
@@ -0,0 +1,22 @@
+            "skylineBranchName": "develop",
+            "skylineS3BranchDirName": "develop",
+            "fromBuildJobUrl": process.env.TRAVIS_BUILD_WEB_URL,
+            "sourceRepo": process.env.TRAVIS_REPO_SLUG
```
we should figure out if we want to use env vars here, or make these function args
I am inclined to make them function args, so that we can test this in isolation
the call can supply the args from the env vars instead

addisonm replied:
I'll make it arguments since I'm doing it for other things.  The isolation testing is a good byproduct as well and even allows potential for manual invocation

rblaa commented on src/deploy/jenkins-uat.js:
```diff
@@ -0,0 +1,22 @@
+            "sourceRepo": process.env.TRAVIS_REPO_SLUG
+        })
+    }).catch(error => {
+        console.log(error);
```
let's put a bit on context so that the error message is clear:
```
console.error('error invoking Jenkins automation at ' + jenkinsUrl + '\n. ' +error);

rblaa commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "truex-shared",
-  "version": "1.6.13",
+  "version": "1.7.1",
```
@addisonm can you use the 1.7.3 version and tag going forward
i need to make another intervening truex-shared fix for automation that i hope to merge today
so i will be taking the 1.7.2 tag
you will then need to merge that in to your branch first so that the final tags are cumulative

addisonm committed:
Merge branch 'develop' of https://github.com/socialvibe/truex-shared-js into CTV-3359-UAT-Automation

# Conflicts:
#	CHANGELOG.md
#	package.json

addisonm committed:
CTV-3359
- Parameterize and rename the jenkins automation
- Clean up

addisonm committed:
CTV-3359
- Fix typo

addisonm committed:
CTV-3359
- Re-order for a bit more consistency

rblaa commented on CHANGELOG.md:
```diff
@@ -1,8 +1,11 @@
 # Changelog
 
-## v1.7.1
+## v1.7.3
```
should be 1.7.4

addisonm replied:
I'll need to do 1.7.5 with how pushed the tags to test it on travis.

rblaa commented on src/deploy/invokeAutomationTests.js:
```diff
@@ -0,0 +1,25 @@
+            "devicePlatforms": devicePlatforms || "Unknown,Tizen,LG,XboxOne,FireTV,AndroidTV,Vizio,Comcast",
+            "whakapapaBranchName": whakapapaBranchName,
+            "skylineBranchName": skylineBranchName,
+            "skylineS3BranchDirName": skylineS3BranchDirName,
```
we can default this if missing, to be computed from the skylineBranchName

addisonm replied:
This looked more appropriate to be based off the pull request rather than skyline branch.  Since this is for the test branches no?  I'll move the guard here instead of from Skyline

rblaa commented on src/deploy/invokeAutomationTests.js:
```diff
@@ -0,0 +1,25 @@
+        },
+        body: JSON.stringify({
+            "devicePlatforms": devicePlatforms || "Unknown,Tizen,LG,XboxOne,FireTV,AndroidTV,Vizio,Comcast",
+            "whakapapaBranchName": whakapapaBranchName,
```
default to develop if missing
is this actually needed for the caller to know?

rblaa replied:
can you file a ticket for wakapapa to support an "All" value so we don't have to maintain the known set here
e.g. what happens when we can start including PS4 and PS5?

addisonm committed:
CTV-3359
- Clean up to be more consistent naming
- Hard code whakapapa branch to be develop

addisonm committed:
CTV-3359
- Fix a filetype issue causing dependency issues

addisonm committed:
CTV-3359
- Fix an issue where an object export is not necessary

addisonm committed:
CTV-3359
- Update changelog

addisonm committed:
CTV-3359
- Switch to "all" fallback in anticipation of whakapapa improvement
- Move s3 branch name logic to here

rblaa commented on src/deploy/deployHelper.js:
```diff
@@ -0,0 +1,8 @@
+
+const getS3BucketPath = (branch) => {
+    return `branch-test/skyline/${branch.replace(/[^a-z0-9_\-]/gi, "_").toLowerCase()}`;
```
for this to be in trues-shared, the repo needs to be an arg
branch-test is also quite specific, and context-dependent
do this instead, rename the function to getS3BranchFolderName, use the function syntax since it is easier to read:
```
function getS3BranchFolderName(branch) {
  return branch.replace(/[^a-z0-9_\-]/gi, "_").toLowerCase();
}

now if you actually need the full path, then you need the prefix to be passed in
basically it really needs to be the caller, since master or develop for Skyline is different than master or develop for C3, etc.

addisonm committed:
CTV-3359
- Remove fallback logic since caller will decide parts of it.  Also only to be used from skyline for now.

addisonm committed:
Merge branch 'develop' of https://github.com/socialvibe/truex-shared-js into CTV-3359-UAT-Automation

# Conflicts:
#	CHANGELOG.md

imann24 commented on CHANGELOG.md:
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+## v1.7.10
```
Looks like we're missing an update to the version in the `package.json` file

addisonm committed:
CTV-3359
- Update package version

rblaa commented:
travis checks are failing

rblaa reviewed (approved):
should fix the travis failures

addisonm commented:
Noting that there are intermittent travis failures.  One is known, the other (which failed for this PR is not known, but was fixed with a travis re-run)
